File: apps/worker/src/worker/asset_processing/extraction.py
import io
import logging
import uuid
from typing import Tuple

import fitz
import numpy as np
from PIL import Image

from worker.asset_processing.image_quality import is_meaningful_image

log = logging.getLogger(__name__)


def _trace_id() -> str:
    return str(uuid.uuid4())[:8]


def _render_page_pil(page: fitz.Page, scale: float = 2.0) -> Image.Image:
    mat = fitz.Matrix(scale, scale)
    pix = page.get_pixmap(matrix=mat, alpha=False)
    return Image.frombytes("RGB", [pix.width, pix.height], pix.samples)


def _extract_embedded_images(
    page: fitz.Page, page_idx: int
) -> list[Tuple[str, Image.Image]]:
    results = []
    doc = page.parent
    for img_idx, img_info in enumerate(page.get_images(full=True)):
        xref = img_info[0]
        try:
            base_img = doc.extract_image(xref)
            img_bytes = base_img["image"]
            pil = Image.open(io.BytesIO(img_bytes)).convert("RGB")
            label = f"page{page_idx}_embedded_{img_idx}"
            results.append((label, pil))
        except Exception:
            pass
    return results


def _extract_rendered_regions(
    page_pil: Image.Image, page_idx: int, min_region_height: int = 120
) -> list[Tuple[str, Image.Image]]:
    gray = np.array(page_pil.convert("L"))
    row_has_content = (gray < 230).any(axis=1)

    regions = []
    in_block = False
    start = 0
    for i, active in enumerate(row_has_content):
        if active and not in_block:
            in_block, start = True, i
        elif not active and in_block:
            in_block = False
            if (i - start) > min_region_height:
                y1 = max(0, start - 15)
                y2 = min(page_pil.height, i + 15)
                regions.append(
                    (
                        f"page{page_idx}_region_{start}",
                        page_pil.crop((0, y1, page_pil.width, y2)),
                    )
                )
    if in_block and (len(row_has_content) - start) > min_region_height:
        y1 = max(0, start - 15)
        regions.append(
            (
                f"page{page_idx}_region_{start}",
                page_pil.crop((0, y1, page_pil.width, page_pil.height)),
            )
        )
    return regions


def extract_visual_elements(
    page: fitz.Page, page_idx: int
) -> list[Tuple[str, Image.Image]]:
    """
    Step 1: Extract embedded PDF image objects
    Step 2: Render page and extract contiguous non-white blocks
    Step 3: Filter by quality
    """
    trace = _trace_id()
    log.info("[trace_id=%s] Extracting visuals from page %d", trace, page_idx)
    embedded = _extract_embedded_images(page, page_idx)
    page_pil = _render_page_pil(page, scale=2.0)
    rendered = _extract_rendered_regions(page_pil, page_idx)
    all_candidates = embedded + rendered
    result = [(label, img) for label, img in all_candidates if is_meaningful_image(img)]
    log.info("[trace_id=%s] Extracted %d meaningful visuals", trace, len(result))
    return result

File: apps/worker/src/asset-processing/extraction.ts
import sharp from "sharp";
import { isMeaningfulImage } from "./image-quality.js";

export interface ExtractedVisual {
  label: string;
  imageBuffer: Buffer;
}

async function extractRenderedRegions(
  pageBuffer: Buffer,
  pageIdx: number,
  minRegionHeight = 120
): Promise<ExtractedVisual[]> {
  const { data, info } = await sharp(pageBuffer)
    .raw()
    .toBuffer({ resolveWithObject: true });

  const width = info.width;
  const height = info.height;
  const channels = info.channels ?? 3;

  const rowHasContent: boolean[] = [];
  for (let y = 0; y < height; y++) {
    let hasContent = false;
    for (let x = 0; x < width; x++) {
      const idx = (y * width + x) * channels;
      const r = data[idx];
      const g = data[idx + 1];
      const b = data[idx + 2];
      if (r < 230 || g < 230 || b < 230) {
        hasContent = true;
        break;
      }
    }
    rowHasContent.push(hasContent);
  }

  const regions: ExtractedVisual[] = [];
  let inBlock = false;
  let start = 0;

  for (let i = 0; i < rowHasContent.length; i++) {
    const active = rowHasContent[i];
    if (active && !inBlock) {
      inBlock = true;
      start = i;
    } else if (!active && inBlock) {
      inBlock = false;
      if (i - start > minRegionHeight) {
        const y1 = Math.max(0, start - 15);
        const y2 = Math.min(height, i + 15);
        const regionBuffer = await sharp(pageBuffer)
          .extract({ left: 0, top: y1, width, height: y2 - y1 })
          .png()
          .toBuffer();
        regions.push({
          label: `page${pageIdx}_region_${start}`,
          imageBuffer: regionBuffer,
        });
      }
    }
  }

  if (inBlock && rowHasContent.length - start > minRegionHeight) {
    const y1 = Math.max(0, start - 15);
    const regionBuffer = await sharp(pageBuffer)
      .extract({ left: 0, top: y1, width, height: height - y1 })
      .png()
      .toBuffer();
    regions.push({
      label: `page${pageIdx}_region_${start}`,
      imageBuffer: regionBuffer,
    });
  }

  return regions;
}

export async function extractVisualElements(
  pageBuffer: Buffer,
  pageIdx: number
): Promise<ExtractedVisual[]> {
  if (!pageBuffer || pageBuffer.length === 0) return [];

  const candidates = await extractRenderedRegions(pageBuffer, pageIdx);
  const result: ExtractedVisual[] = [];

  for (const { label, imageBuffer } of candidates) {
    const meaningful = await isMeaningfulImage(imageBuffer);
    if (meaningful) {
      result.push({ label, imageBuffer });
    }
  }

  return result;
}
